### PR TITLE
Code cleanup and enum performance enhancements

### PR DESCRIPTION
**Pull request by DSupuran**
I'm achieving a 10-15% performance increase when running AngleSharp.Performance.Selector.  If you have a chance please pull these changes and test them as well to make sure I haven't lost my sanity.  All tests are passing in this commit

It looks like I got a little ambitious with changing NodeFlags.cs, so I left that out pending further discussion with you at a later time, but the other changes are included here.

### CODE REVIEW

File: src/AngleSharp.Core.Tests/External/PageRequester.cs [hunk 1]
```diff
@@ -51,8 +51,8 @@ Task<IResponse> ReadResourceAsync(String url)
         async Task<String> AddResourceAsync(String url, IResponse response)
         {
             var counter = 1;
-            var file = default(String);
-            var path = default(String);
+            string file;
```
Comment (FlorianRappl) [inline src/AngleSharp.Core.Tests/External/PageRequester.cs]: Please revert. The coding style of this project uses the RHS convention.

File: src/AngleSharp.Core.Tests/Html/HtmlTokenization.cs [hunk 2]
```diff
@@ -139,7 +139,7 @@ public void TokenizationCharacterReferenceNotin()
             var src = "I'm &notin; I tell you";
             var s = new TextSource(src);
             var t = CreateTokenizer(s);
-            var token = default(HtmlToken);
+            HtmlToken token;
```
Comment (FlorianRappl) [inline src/AngleSharp.Core.Tests/Html/HtmlTokenization.cs]: See comment about RHS.

File: src/AngleSharp.Core.Tests/Html/HtmlTokenization.cs [hunk 3]
```diff
@@ -160,7 +160,7 @@ public void TokenizationCharacterReferenceNotIt()
             var src = "I'm &notit; I tell you";
             var s = new TextSource(src);
             var t = CreateTokenizer(s);
-            var token = default(HtmlToken);
+            HtmlToken token;
```
Comment (FlorianRappl) [inline src/AngleSharp.Core.Tests/Html/HtmlTokenization.cs]: See comment about RHS.

File: src/AngleSharp.Core.Tests/Html/HtmlTokenization.cs [hunk 4]
```diff
@@ -208,7 +208,7 @@ public void TokenizationCDataCorrectCharacters()
             StringBuilder sb = new StringBuilder();
             var s = new TextSource("<![CDATA[hi mum how <!-- are you doing />]]>");
             var t = CreateTokenizer(s);
-            var token = default(HtmlToken);
+            HtmlToken token;
```
Comment (FlorianRappl) [inline src/AngleSharp.Core.Tests/Html/HtmlTokenization.cs]: See comment about RHS.

File: src/AngleSharp.Core.Tests/Html/Validation.cs [hunk 5]
```diff
@@ -47,7 +47,7 @@ public void BasicValidFormValidation()
   <output></output>
 </form>");
             Assert.AreEqual(1, doc.Forms.Length);
-            var form = doc.Forms[0] as IHtmlFormElement;
+            var form = doc.Forms[0];
```
Comment (FlorianRappl) [inline src/AngleSharp.Core.Tests/Html/Validation.cs:50]: This is not the same check. See the `Assert.IsNotNull` below. Please revert.

File: src/AngleSharp.Core.Tests/Html/Validation.cs [hunk 6]
```diff
@@ -28,7 +28,7 @@ public void BasicInvalidFormValidation()
   <output id=""test6""></output>
 </form>");
             Assert.AreEqual(1, doc.Forms.Length);
-            var form = doc.Forms[0] as IHtmlFormElement;
```
Comment (FlorianRappl) [inline src/AngleSharp.Core.Tests/Html/Validation.cs:31]: This is not the same check. See the `Assert.IsNotNull` below. Please revert.

File: src/AngleSharp.Core.Tests/Css/HtmlCssIntegration.cs [hunk 7]
```diff
@@ -42,7 +42,7 @@ public void DetectStylesheet()
             var style = css.Rules[0] as CssStyleRule;
             Assert.AreEqual("body", style.SelectorText);
             Assert.AreEqual(1, style.Style.Length);
-            var decl = style.Style as CssStyleDeclaration;
```
Comment (FlorianRappl) [inline src/AngleSharp.Core.Tests/Css/HtmlCssIntegration.cs:45]: This is not the same check. See the `Assert.IsNotNull` below. Please revert.

File: src/AngleSharp/Dom/Collections/NamedNodeMap.cs [hunk 8]
```diff
@@ -60,7 +60,7 @@ internal void FastAddItem(Attr attr)
 
         internal void RaiseChangedEvent(Attr attr, String newValue, String oldValue)
         {
-            var element = default(Element);
+            Element element;
```
Comment (FlorianRappl) [inline src/AngleSharp/Dom/Collections/NamedNodeMap.cs]: RHS (further violations won't be commented).

File: src/AngleSharp/TextEncoding.cs [hunk 9]
```diff
@@ -317,82 +317,256 @@ static Encoding GetEncoding(String name)
 
         static Dictionary<String, Encoding> CreateEncodings()
         {
-            var encodings = new Dictionary<String, Encoding>(StringComparer.OrdinalIgnoreCase);
+            var encodings = new Dictionary<String, Encoding>(StringComparer.OrdinalIgnoreCase)
+                                {
```
Comment (FlorianRappl) [inline src/AngleSharp/TextEncoding.cs:321]: Make it style consistent, i.e., always use 1-line per entry.

**Comment by FlorianRappl** [comment]
I looked over the commits. Overall, quite alright, however, the RHS convention should not be broken.

The other thing I remarked was that some tests have been altered - the `as` cast was actually crucial as the `Assert.IsNotNull` does not make much sense without them. One may argue that an `Assert.IsInstanceOf<T>` may be better than the `as T`, but in the current version none is realized.

Could you have another look at that?

Much appreciated so far! 👍 👍 👍

**Comment by jogibear9988** [comment]
whats the RHS convention?

**Comment by FlorianRappl** [comment]
That types are only expressed on the right hand side of an (assignment) expression.

**Comment by DSupuran** [comment]
I resubmitted (to what I can see) all the requested changes.  I'm not too
familiar with the pull request on GitHub.  Should I submit another pull
request or does it keep an open history to the commit history of the forked
branch still?

**Comment by FlorianRappl** [comment]
Nope you don't have to perform another PR - as the PR is still open it gets updated with your recently pushed commits.

Looks very good - I'll proceed with the merge! Thanks again, much appreciated 👍 ✌️

**Pull request merged**